### DISCUSSION TITLE
Export references to shared objects

### DISCUSSION THREAD

**Discussion by LKAMinco** [StjerneIdioten/I3D-Blender-Addon#190] [Ideas & Feature Requests]
Add reference objects, which can be exported.

Transform group has attribute _**referenceId**_ which refers to file with i3d
e.g.  
`<File fileId="4" filename="$data/shared/assets/lights/hella/hellaWorkingLightRound_01.i3d" />`
`<TransformGroup name="hellaWorkingLightRound_01_16"  referenceId="4" nodeId="27"/>`
Transform group have also custom user attributes for the reference

```
<UserAttribute nodeId="27">
      <Attribute name="i3d_referencedLightPath" type="string" value="bin/data/shared/assets/lights/hella/hellaWorkingLightRound_01.ma"/>
      <Attribute name="i3d_refModel" type="string" value="hellaWorkingLightRound_01"/>
      <Attribute name="i3d_refBrand" type="string" value="hella"/>
      <Attribute name="i3d_locatorPrefix" type="string" value="locatorRef_"/>
 </UserAttribute>
```

**Comment by StjerneIdioten** [comment]
> `<File fileId="4" filename="$data/shared/assets/lights/hella/hellaWorkingLightRound_01.i3d" />`
> `<TransformGroup name="hellaWorkingLightRound_01_16" referenceId="4" nodeId="27"/>` 

This part requires adding some new stuff to both the UI and i3d generation, but the "user facing" part is pretty straight forward I think.

> ```
> <UserAttribute nodeId="27">
>       <Attribute name="i3d_referencedLightPath" type="string" value="bin/data/shared/assets/lights/hella/hellaWorkingLightRound_01.ma"/>
>       <Attribute name="i3d_refModel" type="string" value="hellaWorkingLightRound_01"/>
>       <Attribute name="i3d_refBrand" type="string" value="hella"/>
>       <Attribute name="i3d_locatorPrefix" type="string" value="locatorRef_"/>
>  </UserAttribute>
> ```

This on the other hand I think requires a bit more consideration for how it should fit into the UI. The exporter already supports custom user attributes, so in theory you could just add the previous part and let people configure this themselves. But it doesn't seem terribly user friendly 😆 I don't know how you are supposed to figure out those values though, just looking at in-game vehicles?

**Reply by StjerneIdioten** [reply]
I can also see it refers to some `.ma` file, which I would assume, given that GE uses Maya, that it is just the Maya Ascii format and the rest of the stuff is keys to load some stuff from that?

**Reply by LKAMinco** [reply]
```
      <Attribute name="i3d_referencedLightPath" type="string"  value="bin/data/shared/assets/lights/hella/hellaWorkingLightRound_01.ma"/> -> seems like path to the file with bin prefix and .ma postfix instead of .i3d
      <Attribute name="i3d_refModel" type="string" value="hellaWorkingLightRound_01"/> -> file name
      <Attribute name="i3d_refBrand" type="string" value="hella"/> -> parent folder of file
      <Attribute name="i3d_locatorPrefix" type="string" value="locatorRef_"/> -> seems static
```

**Reply by StjerneIdioten** [reply]
Do you use this for anything other than lights? Is it always those 4 user attributes or can it be different ones?

**Reply by LKAMinco** [reply]
it can be camera, reflector, etc, probably anything from assets or i3d

**Comment by StjerneIdioten** [comment]
> it can be camera, reflector, etc, probably anything from assets or i3d

So you would always have to manually input it and it can't just be deduced from the loaded i3d file or some other data somehow?

**Reply by LKAMinco** [reply]
Ohh, u don't specify what is it, it is just an model, so all what u need to do, is load path to the i3d file. What i did not tested, is what happens when there are more then one root objecs in reference file

**Reply by StjerneIdioten** [reply]
> Ohh, u don't specify what is it, it is just an model, so all what u need to do, is load path to the i3d file. What i did not tested, is what happens when there are more then one root objecs in reference file

I am talking for the user attributes. Some of theme are clearly named based on the fact that what you are loading is an i3d file with a light in it. Like "i3d_referencedLightPath".

**Comment by StjerneIdioten** [comment]
8b260c3 implements this and only adds the reference functionality. If you need any of the user attributes they can be added manually already 😊
